### PR TITLE
Add alternatebiddercodes in Prebid request ext

### PR DESCRIPTION
**Pull request by pm-nilesh-chate**
w.r.t https://github.com/prebid/prebid-server/issues/2174#issuecomment-1190736031

This changes would make alternatebiddercodes from PBS config formally part of Prebid ORTB extensions.

- Incoming ORTB would contain:
```
ext.prebid.alternatebiddercodes.enabled: true/false
ext.prebid.alternatebiddercodes.bidders.BIDDER.enabled: true/false
ext.prebid.alternatebiddercodes.bidders.BIDDER.allowedbiddercodes: [array, of, strings]
```
- This would override the account-level config
- If the request doesn't specify these values, then PBS-core would merge the account config into these values
- Before sending to each adapter, PBS-core would need to remove entries that don't belong to that bidder. i.e. they would only receive ext.prebid.alternatebiddercodes.adapters.ME
- Everything else is up to the adapter. They can ignore or use at will.

Ex. s2s config on page and incoming PBS request
```
pbjs.setConfig({
    s2sConfig: {
        extPrebid: {
            "alternatebiddercodes": {
                "enabled": true,
                "bidders": {
                    "pubmatic": {
                        "enabled": true,
                        "allowedbiddercodes": [
                            "groupm"
                        ]
                    }
                }
            }
        }
    }
});
```

```
{
   ...
   "ext": {
       "prebid": {
           "alternatebiddercodes": {
                "enabled": true,
                "bidders": {
                    "pubmatic": {
                        "enabled": true,
                        "allowedbiddercodes": [
                            "groupm"
                        ]
                    }
                }
            }
       }
   }
}
```

More details at https://github.com/prebid/prebid-server/issues/2174

### CODE REVIEW

File: exchange/utils.go [hunk 1]
```diff
@@ -297,13 +297,27 @@ func getBidderParamsForBidder(bidderParamsInReqExt map[string]map[string]json.Ra
...
 	if len(req.Ext) == 0 || unpackedExt == nil {
 		return json.RawMessage(``), nil
 	}
 
 	extCopy := *unpackedExt
 	extCopy.Prebid.SChains = nil
+	extCopy.Prebid.AlternateBidderCodes = nil
+
+	//fill request.ext.prebid.alternatebiddercode to have data only of the current bidder
+	if unpackedExt.Prebid.AlternateBidderCodes != nil && len(unpackedExt.Prebid.AlternateBidderCodes.Bidders) != 0 {
```
Comment (pm-nilesh-chate) [inline exchange/utils.go:310]: Do we want to pass `openrtb_ext.ExtAlternateBidderCodes{Enabled: true/false}` to the bidders that are not present in the list req.ext?.

Current changes do not pass it as I don't find any use case for the same.

**Comment by VeronikaSolovei9** [comment]
Code looks good, local testing looks good too.

File: config/accounts.go [hunk 2]
```diff
@@ -18,20 +16,22 @@ const (
...
+	CacheTTL                DefaultTTLs           `mapstructure:"cache_ttl" json:"cache_ttl"`
+	EventsEnabled           bool                  `mapstructure:"events_enabled" json:"events_enabled"`
+	CCPA                    AccountCCPA           `mapstructure:"ccpa" json:"ccpa"`
+	GDPR                    AccountGDPR           `mapstructure:"gdpr" json:"gdpr"`
+	DebugAllow              bool                  `mapstructure:"debug_allow" json:"debug_allow"`
+	DefaultIntegration      string                `mapstructure:"default_integration" json:"default_integration"`
+	CookieSync              CookieSync            `mapstructure:"cookie_sync" json:"cookie_sync"`
+	Events                  Events                `mapstructure:"events" json:"events"` // Don't enable this feature. It is still under developmment - https://github.com/prebid/prebid-server/issues/1725
+	TruncateTargetAttribute *int                  `mapstructure:"truncate_target_attr" json:"truncate_target_attr"`
+	AlternateBidderCodes    *AlternateBidderCodes `mapstructure:"alternatebiddercodes" json:"alternatebiddercodes"`
```
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply config/accounts.go]: Reverted the change. Yes, I did it to detect the presence of account config values but I guess we don't need it.

btw, we would need the AlternateBidderCodes in the request.ext to be pointer to detect its presence and to choose between request and account values.

File: exchange/utils.go [hunk 3]
```diff
@@ -831,3 +845,15 @@ func WrapJSONInData(data []byte) []byte {
 	res = append(res, []byte(`}`)...)
 	return res
 }
+
+// getExtAlternateBidderCodes map config.account.alternatebiddercodes to ext.prebid.alternatebiddercodes
```
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply exchange/utils.go]: done

File: exchange/utils.go [hunk 4]
```diff
@@ -831,3 +845,15 @@ func WrapJSONInData(data []byte) []byte {
...
+func setExtAlternateBidderCodes(requestExt *openrtb_ext.ExtRequest, cfgABC *config.AlternateBidderCodes) {
```
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply exchange/utils.go]: done

File: exchange/exchange.go [hunk 5]
```diff
@@ -494,7 +496,7 @@ func (e *exchange) getAllBids(
 	accountDebugAllowed bool,
 	globalPrivacyControlHeader string,
 	headerDebugAllowed bool,
-	alternateBidderCodes config.AlternateBidderCodes,
+	alternateBidderCodes *openrtb_ext.ExtAlternateBidderCodes,
```
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply exchange/exchange.go]: done

File: exchange/exchange.go [hunk 6]
```diff
@@ -245,6 +245,8 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
 		}
 	}
 
+	setExtAlternateBidderCodes(requestExt, r.Account.AlternateBidderCodes)
```
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply exchange/exchange.go]: sure, moved `setExtAlternateBidderCodes`'s code into `CleanOpenRTBRequests`'s `buildRequestExtForBidder`.

It looks messy now, too many scenarios, although tried to cover all them in unit tests.

File: exchange/utils.go [hunk 7]
```diff
@@ -831,3 +845,15 @@ func WrapJSONInData(data []byte) []byte {
...
+	if requestExt == nil || requestExt.Prebid.AlternateBidderCodes != nil {
```
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply exchange/utils.go]: missed this one. Updated the code in `buildRequestExtForBidder` and added unit test for the same.

File: exchange/exchangetest/extra-bids.json [hunk 8]
```diff
@@ -40,6 +40,15 @@
...
             "includebidderkeys": true
+          },
+          "alternatebiddercodes": {
+            "enabled": true,
+            "bidders": {
+              "pubmatic": {
+                "enabled": true,
+                "allowedbiddercodes": ["groupm"]
+              }
+            }
```
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply exchange/exchangetest/extra-bids.json:51]: Agreed, also I could not test end to end flow at one place where we could validate outgoing and incoming request as these json file based tests mock `requestBid` function which has alternatebiddercodes validation.

Hence, I wrote `TestOverrideConfigAlternateBidderCodesWithRequestValues` to cover the validation of alternatebiddercodes coming as response from adapters and added unit tests in all the possible functions that could change `request.ext.prebid. alternatebiddercodes` to cover the request to adapters part. Ex. `setExtAlternateBidderCodes` and `CleanOpenRTBRequests` (now clubbed).

**Comment by pm-nilesh-chate** [comment]
Hi @bsardo, I've address the comments. Could you review the same.

File: exchange/exchangetest/extra-bids.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply exchange/exchangetest/extra-bids.json:51]: added `expectRequest`

File: exchange/utils.go [hunk 9]
```diff
@@ -48,6 +49,7 @@ func cleanOpenRTBRequests(ctx context.Context,
 	gdprPermsBuilder gdpr.PermissionsBuilder,
 	tcf2ConfigBuilder gdpr.TCF2ConfigBuilder,
 	hostSChainNode *openrtb2.SupplyChainNode,
+	cfgABC config.AlternateBidderCodes,
```
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply exchange/utils.go]: done

File: exchange/utils.go [hunk 10]
```diff
@@ -273,9 +276,36 @@ func getAuctionBidderRequests(auctionRequest AuctionRequest,
...
+	if len(requestExt) == 0 || requestExtParsed == nil || requestExtParsed.Prebid.AlternateBidderCodes == nil {
+		alternateBidderCodes = &openrtb_ext.ExtAlternateBidderCodes{
+			Enabled: cfgABC.Enabled,
+		}
+		if bidderCodes, ok := cfgABC.Bidders[bidder]; ok {
+			alternateBidderCodes = &openrtb_ext.ExtAlternateBidderCodes{
+				Bidders: map[string]openrtb_ext.ExtAdapterAlternateBidderCodes{
+					bidder: bidderCodes,
+				},
+			}
```
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply exchange/utils.go]: done. Added similar test case for `buildRequestExtForBidder()`

File: exchange/utils.go [hunk 11]
```diff
@@ -273,9 +276,36 @@ func getAuctionBidderRequests(auctionRequest AuctionRequest,
...
+		}
+
+		// default case where alternatebiddercode is either not defined in account config or defined with default values,
+		// do not fill alternatebiddercode in such scenario to keep the request payload as is (regression)
+		if reflect.DeepEqual(openrtb_ext.ExtAlternateBidderCodes{}, *alternateBidderCodes) {
+			if len(requestExt) == 0 || requestExtParsed == nil {
+				return json.RawMessage(``), nil
+			}
+			alternateBidderCodes = nil
+		}
```
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply exchange/utils.go]: done

File: exchange/utils.go [hunk 12]
```diff
@@ -273,9 +276,36 @@ func getAuctionBidderRequests(auctionRequest AuctionRequest,
 	return bidderRequests, errs
 }
 
-func buildRequestExtForBidder(bidder string, requestExt json.RawMessage, requestExtParsed *openrtb_ext.ExtRequest, bidderParamsInReqExt map[string]json.RawMessage) (json.RawMessage, error) {
-	if len(requestExt) == 0 || requestExtParsed == nil {
-		return json.RawMessage(``), nil
+func buildRequestExtForBidder(bidder string, requestExt json.RawMessage, requestExtParsed *openrtb_ext.ExtRequest, bidderParamsInReqExt map[string]json.RawMessage, cfgABC config.AlternateBidderCodes) (json.RawMessage, error) {
+	// Resolve alternatebiddercode for current bidder
```
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply exchange/utils.go:277]: done

**Comment by pm-nilesh-chate** [comment]
@bsardo I've address the comments. Could you review the same.

Thank you for the critical inputs. A good learning for my future PRs :)

**Review by VeronikaSolovei9 — APPROVED** [review]
Code looks so much better and easier to understand now. LGTM

File: exchange/utils_test.go [hunk 13]
```diff
@@ -3272,3 +3302,159 @@ func TestBuildExtData(t *testing.T) {
...
+			},
+			wantError: false,
+		},
+		{
+			desc:     "Nil request ext, alternatebiddercodes config enabled but bidder not present",
+			inExt:    nil,
+			inCfgABC: config.AlternateBidderCodes{Enabled: true},
+			wantExt: []json.RawMessage{
+				json.RawMessage(`{"prebid":{"alternatebiddercodes":{"enabled":true,"bidders":null}}}`),
+				json.RawMessage(`{"prebid":{"alternatebiddercodes":{"enabled":true,"bidders":null}}}`),
```
Comment (bsardo) [inline exchange/utils_test.go:3360]: I see a few instances where `prebid.alternatebiddercodes.bidders` is `null`. Perhaps this should be omitted if null, in which case we would add the `omitempty` tag to the `ExtAlternateBidderCodes.Bidders` field. Thoughts?

File: exchange/utils.go [hunk 14]
```diff
@@ -302,14 +313,42 @@ func buildRequestExtForBidder(bidder string, requestExt json.RawMessage, request
...
 
 	return json.Marshal(extMap)
 }
 
+func buildRequestExtAlternateBidderCodes(bidder string, accABC config.AlternateBidderCodes, reqABC *openrtb_ext.ExtAlternateBidderCodes) *openrtb_ext.ExtAlternateBidderCodes {
+	if reqABC != nil {
+		return &openrtb_ext.ExtAlternateBidderCodes{
+			Enabled: reqABC.Enabled,
+			Bidders: map[string]openrtb_ext.ExtAdapterAlternateBidderCodes{
+				bidder: reqABC.Bidders[bidder],
```
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply exchange/utils.go]: done. Also added a test case for this scenario

File: exchange/exchange.go [hunk 15]
```diff
@@ -286,7 +286,14 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
 	} else {
 		// List of bidders we have requests for.
 		liveAdapters = listBiddersWithRequests(bidderRequests)
-		adapterBids, adapterExtra, anyBidsReturned = e.getAllBids(auctionCtx, bidderRequests, bidAdjustmentFactors, conversions, accountDebugAllow, r.GlobalPrivacyControlHeader, debugLog.DebugOverride, r.Account.AlternateBidderCodes, requestExt.Prebid.Experiment)
+
+		//This will be used validate bids
+		alternateBidderCodes := openrtb_ext.ExtAlternateBidderCodes(r.Account.AlternateBidderCodes)
+		if requestExt != nil && requestExt.Prebid.AlternateBidderCodes != nil {
+			alternateBidderCodes = *requestExt.Prebid.AlternateBidderCodes
+		}
```
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply exchange/exchange.go:294]: Added this to make sure PBS-Core data is used to validate the adapter's response instead of relying on the data from adapter's request which could be altered by the adapter code.

File: exchange/utils_test.go (hunk 13 above)
Comment (pm-nilesh-chate) [reply exchange/utils_test.go]: Added omit empty to all the alternatebiddercodes fields.

The only downside I see here is that the SSPs would not know whether these fields were omitted by the user or by the PBS. One will have to refer the incoming payload to PBS to verify the same. I don't see any harm though.

**Comment by pm-nilesh-chate** [comment]
Addressed the review comments. Had to rebase and force push to fix the master merge test failures.

@bsardo Could you please review the latest changes.

Thanks

**Comment by abhinavsinha001** [comment]
What is the holdup here can we get these changes merged? 
We want to start using this functionality.

**Comment by bsardo** [comment]
@abhinavsinha001 I'm waiting for a response on this https://github.com/prebid/prebid-server/issues/2174#issuecomment-1242045987

**Comment by abhinavsinha001** [comment]
Thanks @bsardo for pointing to the comment.
@bretg can you please help unblock here.

File: exchange/utils.go [hunk 16]
```diff
@@ -302,14 +313,45 @@ func buildRequestExtForBidder(bidder string, requestExt json.RawMessage, request
...
+		if err := json.Unmarshal(requestExt, &extMap); err != nil {
+			return nil, err
+		}
 	}
 	extMap["prebid"] = prebidJson
 
 	return json.Marshal(extMap)
 }
 
+func buildRequestExtAlternateBidderCodes(bidder string, accABC config.AlternateBidderCodes, reqABC *openrtb_ext.ExtAlternateBidderCodes) *openrtb_ext.ExtAlternateBidderCodes {
```
Comment (bsardo) [inline exchange/utils.go:326]: After looking more closely at the requirements, I believe we should aim for the following behavior:
- If `alternatebiddercodes` is not defined in the account or on the request, the field should be omitted from all bidder requests.
- If `alternatebiddercodes` is defined either in the account or on the request but its `enabled` is not set, the `alternatebiddercodes` field should be added to the bidder request with the default value of false.
- The `alternatebiddercodes` object set on a bidder request should only contain its own bidder object in the `bidders` field and it should match the bidder object in the account/request it was sourced from. Note an undefined map or slice in the original request or account will appear as `null` in a bidder request (we will need to get rid of `omitempty`), so that we can distinguish between an empty array/map and an undefined array/map.

Given the above, I'd like to suggest a couple of things here:
1) In `config/accounts.go`, can we make the field `AlternateBidderCodes` in the `Account` struct of type `*AlternateBidderCodes`?
2) Let's get rid of the `if` block on line 348 that sets `alternateBidderCodes = nil`.
3) I think this function ultimately looks like this:
```
        if reqABC != nil {
		alternateBidderCodes := &openrtb_ext.ExtAlternateBidderCodes{
			Enabled: reqABC.Enabled,
		}
		if bidderCodes, ok := reqABC.Bidders[bidder]; ok {
			alternateBidderCodes.Bidders = map[string]openrtb_ext.ExtAdapterAlternateBidderCodes{
				bidder: bidderCodes,
			}
		}
		return alternateBidderCodes
	}

	if accABC != nil {
		alternateBidderCodes := &openrtb_ext.ExtAlternateBidderCodes{
			Enabled: accABC.Enabled,
		}
		if bidderCodes, ok := accABC.Bidders[bidder]; ok {
			alternateBidderCodes.Bidders = map[string]openrtb_ext.ExtAdapterAlternateBidderCodes{
				bidder: bidderCodes,
			}
		}
		return alternateBidderCodes
	}
	return nil
```

File: exchange/exchange.go [hunk 17]
```diff
@@ -286,7 +286,14 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
 	} else {
 		// List of bidders we have requests for.
 		liveAdapters = listBiddersWithRequests(bidderRequests)
-		adapterBids, adapterExtra, anyBidsReturned = e.getAllBids(auctionCtx, bidderRequests, bidAdjustmentFactors, conversions, accountDebugAllow, r.GlobalPrivacyControlHeader, debugLog.DebugOverride, r.Account.AlternateBidderCodes, requestExt.Prebid.Experiment)
+
+		//This will be used validate bids
+		alternateBidderCodes := openrtb_ext.ExtAlternateBidderCodes(r.Account.AlternateBidderCodes)
```
Comment (bsardo) [inline exchange/exchange.go:291]: See my related comment in `exchange/utils.go` that asks for the account field to be converted to a pointer.
Should this be something like this now:
```
var alternateBidderCodes openrtb_ext.ExtAlternateBidderCodes
if r.Account.AlternateBidderCodes != nil {
    alternateBidderCodes = openrtb_ext.ExtAlternateBidderCodes(*r.Account.AlternateBidderCodes)
}
```

File: exchange/utils.go [hunk 18]
```diff
@@ -302,14 +313,45 @@ func buildRequestExtForBidder(bidder string, requestExt json.RawMessage, request
...
+	alternateBidderCodes := &openrtb_ext.ExtAlternateBidderCodes{
+		Enabled: accABC.Enabled,
+	}
+	if bidderCodes, ok := accABC.Bidders[bidder]; ok {
+		alternateBidderCodes.Bidders = map[string]openrtb_ext.ExtAdapterAlternateBidderCodes{
+			bidder: bidderCodes,
+		}
+	}
+
+	if !alternateBidderCodes.Enabled && alternateBidderCodes.Bidders == nil {
```
Comment (bsardo) [inline exchange/utils.go:348]: I'm wondering now if we should delete this if block. If we remove it, `ext.prebid.alternatebiddercodes` in the bidder request will contain exactly what's pertinent to the bidder in the account config with the lone exception being that `enabled` flags will be set to the default value false if not defined. This is the same behavior as if `ext.prebid.alternatebiddercodes` were to be populated on the bidder request with information pulled from the original request, assuming we remove the couple of `omitempty` tags I pointed out in my other comments.

**Review by bsardo — COMMENTED** [review]
Thanks for your patience @pm-nilesh-chate. Please see my comments. I ran a bunch of manual tests with my suggested changes to make sure they were in line with my requirements analysis outlined in my comment in `exchange/utils.go`. Let me know if you want to discuss.

File: openrtb_ext/alternatebiddercodes.go [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+// ExtAlternateBidderCodes defines list of alternate bidder codes allowed by adatpers. This overrides host level configs.
+type ExtAlternateBidderCodes struct {
+	Enabled bool                                      `mapstructure:"enabled" json:"enabled"`
+	Bidders map[string]ExtAdapterAlternateBidderCodes `mapstructure:"bidders" json:"bidders,omitempty"`
+}
+
+type ExtAdapterAlternateBidderCodes struct {
+	Enabled            bool     `mapstructure:"enabled" json:"enabled"`
+	AllowedBidderCodes []string `mapstructure:"allowedbiddercodes" json:"allowedbiddercodes,omitempty"`
```
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply openrtb_ext/alternatebiddercodes.go]: done, removed `omitempty` tag.

File: openrtb_ext/alternatebiddercodes.go [hunk 20]
```diff
@@ -0,0 +1,53 @@
+package openrtb_ext
+
+import "fmt"
+
+// ExtAlternateBidderCodes defines list of alternate bidder codes allowed by adatpers. This overrides host level configs.
+type ExtAlternateBidderCodes struct {
+	Enabled bool                                      `mapstructure:"enabled" json:"enabled"`
+	Bidders map[string]ExtAdapterAlternateBidderCodes `mapstructure:"bidders" json:"bidders,omitempty"`
```
[the remark this answers is not in the document]
Comment (pm-nilesh-chate) [reply openrtb_ext/alternatebiddercodes.go]: done, removed `omitempty` tag.

File: exchange/utils.go (hunk 16 above)
Comment (pm-nilesh-chate) [reply exchange/utils.go]: done. made `accABC` pointer to detect its presence.

File: exchange/exchange.go (hunk 17 above)
Comment (pm-nilesh-chate) [reply exchange/exchange.go]: done

File: exchange/utils.go (hunk 18 above)
Comment (pm-nilesh-chate) [reply exchange/utils.go]: agreed. removed this code.

**Comment by pm-nilesh-chate** [comment]
Hi @bsardo, I've addressed the comments. Please review the same. Fyi, last commit (master merge) is to address Trivy's check.

**Review by bsardo — DISMISSED** [review]
LGTM

File: openrtb_ext/alternatebiddercodes.go [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+type ExtAdapterAlternateBidderCodes struct {
+	Enabled            bool     `mapstructure:"enabled" json:"enabled"`
+	AllowedBidderCodes []string `mapstructure:"allowedbiddercodes" json:"allowedbiddercodes"`
+}
+
+func (bidderCodes *ExtAlternateBidderCodes) IsValidBidderCode(bidder, alternateBidder string) (bool, error) {
+	const ErrAlternateBidderNotDefined = "alternateBidderCodes not defined for adapter %q, rejecting bids for %q"
+	const ErrAlternateBidderDisabled = "alternateBidderCodes disabled for %q, rejecting bids for %q"
```
Comment (SyntaxNode) [inline openrtb_ext/alternatebiddercodes.go:18]: The naming convention `errSomeName` is usually for when the value is an error type. This is an error message and the separation of the fields from the Errorf calls is a bit hard to follow IMHO. Consider extracting each to their own simple builder method, like:

```
func alternateBidderNotDefinedError(bidder, alternateBidder string) error {
  return fmt.Errorf("alternateBidderCodes not defined for adapter %q, rejecting bids for %q", bidder, alternateBidder)
}
```

Comment (pm-nilesh-chate) [reply openrtb_ext/alternatebiddercodes.go]: done

File: exchange/exchange.go [hunk 22]
```diff
@@ -286,7 +286,16 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
...
 		// List of bidders we have requests for.
 		liveAdapters = listBiddersWithRequests(bidderRequests)
-		adapterBids, adapterExtra, anyBidsReturned = e.getAllBids(auctionCtx, bidderRequests, bidAdjustmentFactors, conversions, accountDebugAllow, r.GlobalPrivacyControlHeader, debugLog.DebugOverride, r.Account.AlternateBidderCodes, requestExt.Prebid.Experiment)
+
+		//This will be used validate bids
+		var alternateBidderCodes openrtb_ext.ExtAlternateBidderCodes
+		if requestExt != nil && requestExt.Prebid.AlternateBidderCodes != nil {
+			alternateBidderCodes = *requestExt.Prebid.AlternateBidderCodes
+		} else if r.Account.AlternateBidderCodes != nil {
+			alternateBidderCodes = openrtb_ext.ExtAlternateBidderCodes(*r.Account.AlternateBidderCodes)
```
Comment (bsardo) [inline exchange/exchange.go:295]: I think you can remove this cast now since `r.Account.AlternateBidderCodes` is already of type `ExtAlternateBidderCodes`.

**Review by bsardo — APPROVED** [review]

**Review by VeronikaSolovei9 — APPROVED** [review]
LGTM
